### PR TITLE
Add flake, format files and improve readme

### PR DESCRIPTION
**Pull request by deemp**

### CODE REVIEW

**Comment by deemp** [comment]
Hi @patrickt! Started studying `fused-effects` and decided to update some things a bit. Could you please have a look?

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,4 @@ stack.yaml
 cabal.project.local
 .ghci
 public/
+.vscode
```
Comment (patrickt) [inline .gitignore:8]: I think we’d prefer to keep editor-specific stuff out of tree; handling your editor’s ephemera is best left to global .gitignore rules rather than per-project.

File: fourmolu.yaml [hunk 2]
```diff
@@ -0,0 +1 @@
+indentation: 2
```
Comment (patrickt) [inline fourmolu.yaml:1]: I think we’d also prefer to keep beautifiers out of the tree, if only so as to avoid another years-long (and if I exaggerate, it is only slightly) argument between me and @robrix about the virtue (_or lack thereof_) of horizontal alignment. 😄

File: docs/examples/ReinterpretLog.md [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+This example shows how to reinterpret a simple, first-order `logging` effect,
+in terms of itself, in order to change the type of the values it logs.
+
+- First, we will define a structured log message type, which is the type our
+  application prefers to log in.
+
+- Next, we will define a logging carrier that prints strings to `stdout`.
+
+- Finally, we will bridge the two with an effect carrier that reinterprets
+  structured log messages as strings.
```
Comment (patrickt) [inline docs/examples/ReinterpretLog.md:12]: I see from the main changes in the README that this adds the example files as first-class documentation. I think this is a nice idea, as the rendering on this website for a nice Markdown or `.lhs` file will be better. It’s not hugely important to me what form the examples take, whether they’re `.hs` or `.lhs` or `.md`, as long as they render nicely in-browser, but it _is_ important that they not be duplicated in-tree, since that’s how things drift out of sync, and having internally inconsistent docs is worse than no docs at all. So as it stands I would be a no on this change. However, I think you can still thread this particular needle with the `markdown-unlit` tool. Furthermore, removing this duplication will make your Nix story shorter and simpler, since you won’t need the `writeDocs` stage, and I think you can drop your `lima` dependency, if not more.

Here I would suggest copying what we do with the compilable `README.md`: use `-pgmL markdown-unlit`, and specify the thus-created .lhs files in the cabal file.

Alternatively, if `.lhs` files render well on this website, that might be a good choice. If the rendered document is comparable to a native Markdown file I would actually probably be in favor of `.lhs`, unless there’s some other concrete reason to choose Markdown.

File: flake.nix [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  inputs = {
+    nixpkgs_.url = "github:deemp/flakes?dir=source-flake/nixpkgs";
+    nixpkgs.follows = "nixpkgs_/nixpkgs";
+    my-codium.url = "github:deemp/flakes?dir=codium";
+    drv-tools.url = "github:deemp/flakes?dir=drv-tools";
+    flake-utils_.url = "github:deemp/flakes?dir=source-flake/flake-utils";
+    flake-utils.follows = "flake-utils_/flake-utils";
+    haskell-tools.url = "github:deemp/flakes?dir=language-tools/haskell";
+    devshell.url = "github:deemp/flakes?dir=devshell";
+    my-lima.url = "github:deemp/flakes?dir=lima";
```
Comment (patrickt) [inline flake.nix:11]: As mentioned above, I think you can drop this dependency, among others, if you either use .lhs for the docs or .md + markdown-unlit.

File: flake.nix [hunk 5]
```diff
@@ -0,0 +1,137 @@
+{
+  inputs = {
```
Comment (patrickt) [inline flake.nix:2]: So, Nix. Here are some thoughts; hopefully @robrix will chime in.

Nix certainly doesn’t seem to be going anywhere any time soon. If hosting our own Flake file makes it as easy as possible for Nix users to pull `fused-effects` into their project, I’m in favor of it. (If there is a good link showing the difference in workflow for a library with a flake.nix compared to one without one, I’d very much like to see it.) However, neither Rob or I use Nix, so our ability to maintain it is essentially zero. What I would need before accepting this would be:

1. one or more active Nix users to commit to maintaining it (also paging @jkachmar, muahaha); we’d be willing to hand out commit bits for that particular file;
2. CI integration via Actions (otherwise we have no way of knowing when it breaks)
3. a probationary period for a couple of GHC dot-releases: if Nix+Actions can keep up with .cabal in green-ness, then I’d like to keep them, but if they become one of those targets that’s perpetually red, then I’d rather excise it
4. ideally, some documentation on how to ensure version bumps and GHC upgrades go smoothly.

File: flake.nix [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    nixpkgs_.url = "github:deemp/flakes?dir=source-flake/nixpkgs";
+    nixpkgs.follows = "nixpkgs_/nixpkgs";
+    my-codium.url = "github:deemp/flakes?dir=codium";
```
Comment (patrickt) [inline flake.nix:5]: Is there a concrete reason why you would need to have codium in the mix?

File: flake.nix [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+      codium = mkCodium {
+        extensions = { inherit (extensions) nix haskell misc github markdown; };
+        runtimeDependencies = codiumTools;
+      };
+
+      # Write .vscode/settings.json
+      writeSettings = writeSettingsJSON {
+        inherit (settingsNix) haskell todo-tree files editor gitlens
+          git nix-ide workbench markdown-all-in-one markdown-language-features;
+      };
```
Comment (patrickt) [inline flake.nix:68]: As mentioned above, I don’t think our project is the right place to put editor-specific settings, so I’d prefer to see this elided.

**Review by patrickt — CHANGES_REQUESTED** [review]

File: fused-effects.cabal [hunk 8]
```diff
@@ -139,6 +140,13 @@ test-suite examples
     , fused-effects
     , hedgehog           >= 1 && < 1.2
     , hedgehog-fn        ^>= 1
+    , transformers
```
Comment (patrickt) [inline fused-effects.cabal:143]: CI is complaining that this is unnecessary.

**Comment by deemp** [comment]
Hello! Thank you for your review! I should have thought about these things before making a PR...

In a nutshell, this flake should provide a complete IDE with Haskell extensions, shell tools and scripts. This approach might be flawed, but it pretty conveniently isolates projects. As `VSCodium` is my favorite IDE, it's included into the flake. Unfortunately, `VSCodium` requires some configuration via files that are generated from Nix expressions and that should be gitignored (https://github.com/fused-effects/fused-effects/pull/442#discussion_r1072820049). In fact, none of this stuff is necessary unless this project decides to provide such a dev environment. From what I see, this project is fine without Nix.

Now, the examples in `.md`. Basically, there are two ways to obtain a `.md` version of a Haskell program. When using `markdown-unlit`, one writes a `.md` and then extracts `.hs`. On the other hand, when using `lima`, one writes a Haskell program in a `.hs` with multi-line comments in `Markdown`. Then, one converts the `.hs` to `.md`, and comments become text. I prefer the second option as it enables one to use the `Haskell Language Server` (`HLS`) to refactor that `.hs`. That said, I suggest using `lima` for docs generation. If I understand correctly, the `lima` step may replace `markdown-unlit`, then be run in GH Actions and committed there in case `.hs` files with examples change. This way, though we have some duplication, the source code for examples is easy to work with using the standard tools like cabal and HLS. The only prerequisite is to publish `lima` on Hackage.

To sum up, if you don't mind switching to `lima` from `markdown-unlit`, I'll edit the test part in `.cabal` and GH Actions and remove the Nix stuff and IDE stuff. Otherwise, please, close this PR.

**Comment by deemp** [comment]
I removed `Nix` and added tests to generate Markdown via `lima` during `cabal test`

File: README.md [hunk 9]
```diff
@@ -318,7 +287,6 @@ Finally, thanks to the fusion and inlining of carriers, `fused-effects` is only
 
 [`mtl`]: http://hackage.haskell.org/package/mtl
 
-
 #### Comparison to `freer-simple`
```
Comment (robrix) [inline README.md:290]: Double spaces preceding headings are intentional, as they make the document _much_ more readable for me. Please leave them in place.

File: README.md [hunk 10]
```diff
@@ -331,23 +299,20 @@ Finally, `fused-effects` has been [benchmarked](#benchmarks) as faster than `fre
...
 ### Acknowledgements
 
 The authors of fused-effects would like to thank:
 
-* Tom Schrijvers, Nicholas Wu, and all their collaborators for the research that led to `fused-effects`;
-* Alexis King for thoughtful discussions about and suggestions regarding our methodology;
-* the authors of other effect libraries, including `eff`, `polysemy`, and `capabilities`, for their exploration of the space.
+- Tom Schrijvers, Nicholas Wu, and all their collaborators for the research that led to `fused-effects`;
+- Alexis King for thoughtful discussions about and suggestions regarding our methodology;
+- the authors of other effect libraries, including `eff`, `polysemy`, and `capabilities`, for their exploration of the space.
```
Comment (robrix) [inline README.md:318]: 👍🏻 Good catch.

File: docs/examples/ReinterpretLog.md (hunk 3 above)
Comment (robrix) [reply docs/examples/ReinterpretLog.md:12]: It looks like @patrickt's suggestions would need to be applied to the main README.md/README.lhs document as well.

File: examples/ReinterpretLog.hs [hunk 11]
```diff
@@ -21,145 +25,161 @@
...
+-}
 
--- Our structured log message. In this example, we just tag a 'String' with its
--- severity, but this can be anything.
 data Message
   = Debug String
   | Info String
 
--- Render a structured log message as a string.
+{- Render a structured log message as a string. -}
```
Comment (robrix) [inline examples/ReinterpretLog.hs:58]: I am really 👎🏻 on essentially all of the formatting changes in this file. They're needless, they make the diff harder to read, and they specifically do not reflect the formatting used elsewhere in the project.

**Review by robrix — CHANGES_REQUESTED** [review]
Thanks for the PR!

I'm not interested in changing our formatting or any of the workflows around editing the docs (e.g. switching away from `markdown-unlit`). I am definitely open to _discussing_ them, but absent that, I don't think I understand the motivation for the changes in this PR adequately to merge it. If something is broken, definitely let us know, but otherwise I'd prefer to close out this PR. We can reopen if we decide such changes are warranted.

**Comment by robrix** [comment]
Closing per comments in my review above.
